# NYC Airbnb Exploratory Analysis

## Contents
### 1. Import Libraries
### 2. Import Data
### 3. Exploring 
### 4. Data Wrangling
### 5. Data Consistency Check
#### 5a. Missing Data
#### 5b. Duplicates
#### 5c. Data Types
#### 5d. Updating Incorrect Data 
### 6. Export Data

## 1. Importing Libraries

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as ply
import seaborn as sns
import scipy

In [2]:
#set path
path = r'/Users/shelbysteele/Documents/Airbnb NYC Rental Analysis'

## 2. Importing Data

In [5]:
#import Data
df = pd.read_csv(r'/Users/shelbysteele/Documents/Airbnb NYC Rental Analysis/02 Data/Original Data/Airbnb_Open_Data.csv', index_col = False)

/var/folders/4n/57r3q8s10pz5g_h9r3d2bgpw0000gp/T/ipykernel_47545/4039886203.py:2: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(r'/Users/shelbysteele/Documents/Airbnb NYC Rental Analysis/02 Data/Original Data/Airbnb_Open_Data.csv', index_col = False)


## 3. Data Exploration

In [6]:
#checking the data
df.head()

id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1002403               THE VILLAGE OF HARLEM....NEW YORK !  78829239556   
3  1002755                                               NaN  85098326012   
4  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   

  host_identity_verified host name neighbourhood group neighbourhood  \
0            unconfirmed  Madaline            Brooklyn    Kensington   
1               verified     Jenna           Manhattan       Midtown   
2                    NaN     Elise           Manhattan        Harlem   
3            unconfirmed     Garry            Brooklyn  Clinton Hill   
4               verified    Lyndon           Manhattan   East Harlem   

        lat      long        country  ... service fee minimum nights  \
0  40.64749 -73.97237  United States  ...       $193            10.0   
1  40.75362 -73.98377  United States  ...        $28            30.0   
2  40.80902 -73.94190  United States  ...       $124             3.0   
3  40.68514 -73.95976  United States  ...        $74            30.0   
4  40.79851 -73.94399  United States  ...        $41            10.0   

  number of reviews last review  reviews per month review rate number  \
0               9.0    10/19/21               0.21                4.0   
1              45.0     5/21/22               0.38                4.0   
2               0.0         NaN                NaN                5.0   
3             270.0      7/5/19               4.64                4.0   
4               9.0    11/19/18               0.10                3.0   

  calculated host listings count  availability 365  \
0                            6.0             286.0   
1                            2.0             228.0   
2                            1.0             352.0   
3                            1.0             322.0   
4                            1.0             289.0   

                                         house_rules license  
0  Clean up and treat the home the way you'd like...     NaN  
1  Pet friendly but please confirm with me if the...     NaN  
2  I encourage you to use my kitchen, cooking and...     NaN  
3                                                NaN     NaN  
4  Please no smoking in the house, porch or on th...     NaN  

[5 rows x 26 columns]

In [7]:
#checking the shape of the data
df.shape

(102599, 26)

In [8]:
#checking the data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

In [9]:
#checking the columns
df.columns

Index(['id', 'NAME', 'host id', 'host_identity_verified', 'host name',
       'neighbourhood group', 'neighbourhood', 'lat', 'long', 'country',
       'country code', 'instant_bookable', 'cancellation_policy', 'room type',
       'Construction year', 'price', 'service fee', 'minimum nights',
       'number of reviews', 'last review', 'reviews per month',
       'review rate number', 'calculated host listings count',
       'availability 365', 'house_rules', 'license'],
      dtype='object')

In [10]:
#checking the descriptive statistics
df.describe()

id       host id            lat           long  \
count  1.025990e+05  1.025990e+05  102591.000000  102591.000000   
mean   2.914623e+07  4.925411e+10      40.728094     -73.949644   
std    1.625751e+07  2.853900e+10       0.055857       0.049521   
min    1.001254e+06  1.236005e+08      40.499790     -74.249840   
25%    1.508581e+07  2.458333e+10      40.688740     -73.982580   
50%    2.913660e+07  4.911774e+10      40.722290     -73.954440   
75%    4.320120e+07  7.399650e+10      40.762760     -73.932350   
max    5.736742e+07  9.876313e+10      40.916970     -73.705220   

       Construction year  minimum nights  number of reviews  \
count      102385.000000   102190.000000      102416.000000   
mean         2012.487464        8.135845          27.483743   
std             5.765556       30.553781          49.508954   
min          2003.000000    -1223.000000           0.000000   
25%          2007.000000        2.000000           1.000000   
50%          2012.000000        3.000000           7.000000   
75%          2017.000000        5.000000          30.000000   
max          2022.000000     5645.000000        1024.000000   

       reviews per month  review rate number  calculated host listings count  \
count       86720.000000       102273.000000                   102280.000000   
mean            1.374022            3.279106                        7.936605   
std             1.746621            1.284657                       32.218780   
min             0.010000            1.000000                        1.000000   
25%             0.220000            2.000000                        1.000000   
50%             0.740000            3.000000                        1.000000   
75%             2.000000            4.000000                        2.000000   
max            90.000000            5.000000                      332.000000   

       availability 365  
count     102151.000000  
mean         141.133254  
std          135.435024  
min          -10.000000  
25%            3.000000  
50%           96.000000  
75%          269.000000  
max         3677.000000

###### last review has future dates and needs to be removed
###### the minimum is -1223 days while the max is 5645 days
###### the availability for the next 365 daya shows -10 as the min, and 3677 as the max. 

In [11]:
#checking the data types
df.dtypes

id                                  int64
NAME                               object
host id                             int64
host_identity_verified             object
host name                          object
neighbourhood group                object
neighbourhood                      object
lat                               float64
long                              float64
country                            object
country code                       object
instant_bookable                   object
cancellation_policy                object
room type                          object
Construction year                 float64
price                              object
service fee                        object
minimum nights                    float64
number of reviews                 float64
last review                        object
reviews per month                 float64
review rate number                float64
calculated host listings count    float64
availability 365                  

## 4. Data Wrangling

In [12]:
#drop columns for host name due to personal identifiable information (PPI)
df.drop(columns = {'host name'}, inplace = True)

In [13]:
#checking output
df.head()

id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1002403               THE VILLAGE OF HARLEM....NEW YORK !  78829239556   
3  1002755                                               NaN  85098326012   
4  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   

  host_identity_verified neighbourhood group neighbourhood       lat  \
0            unconfirmed            Brooklyn    Kensington  40.64749   
1               verified           Manhattan       Midtown  40.75362   
2                    NaN           Manhattan        Harlem  40.80902   
3            unconfirmed            Brooklyn  Clinton Hill  40.68514   
4               verified           Manhattan   East Harlem  40.79851   

       long        country country code  ... service fee minimum nights  \
0 -73.97237  United States           US  ...       $193            10.0   
1 -73.98377  United States           US  ...        $28            30.0   
2 -73.94190  United States           US  ...       $124             3.0   
3 -73.95976  United States           US  ...        $74            30.0   
4 -73.94399  United States           US  ...        $41            10.0   

  number of reviews  last review reviews per month review rate number  \
0               9.0     10/19/21              0.21                4.0   
1              45.0      5/21/22              0.38                4.0   
2               0.0          NaN               NaN                5.0   
3             270.0       7/5/19              4.64                4.0   
4               9.0     11/19/18              0.10                3.0   

   calculated host listings count  availability 365  \
0                             6.0             286.0   
1                             2.0             228.0   
2                             1.0             352.0   
3                             1.0             322.0   
4                             1.0             289.0   

                                         house_rules  license  
0  Clean up and treat the home the way you'd like...      NaN  
1  Pet friendly but please confirm with me if the...      NaN  
2  I encourage you to use my kitchen, cooking and...      NaN  
3                                                NaN      NaN  
4  Please no smoking in the house, porch or on th...      NaN  

[5 rows x 25 columns]

In [14]:
#correcting column names to be consistent 
df.columns=[col.lower().replace(" ","_") for col in df.columns]
df.columns

Index(['id', 'name', 'host_id', 'host_identity_verified',
       'neighbourhood_group', 'neighbourhood', 'lat', 'long', 'country',
       'country_code', 'instant_bookable', 'cancellation_policy', 'room_type',
       'construction_year', 'price', 'service_fee', 'minimum_nights',
       'number_of_reviews', 'last_review', 'reviews_per_month',
       'review_rate_number', 'calculated_host_listings_count',
       'availability_365', 'house_rules', 'license'],
      dtype='object')

In [15]:
#renaming columns 
df.rename(columns = {'name' : 'listing_title'}, inplace = True)
df.rename(columns = {'neighbourhood_group' : 'borough'}, inplace = True) 
df.rename(columns = {'neighbourhood' : 'neighborhood'}, inplace = True) 
df.rename(columns = {'lat' : 'latitude'}, inplace = True) 
df.rename(columns = {'long' : 'longitude'}, inplace = True) 
df.rename(columns = {'calculated_host_listings_count' : 'listings_per_host'}, inplace = True) 

In [16]:
#replacing brookln with Brooklyn
df['borough'] = df['borough'].replace('brookln', 'Brooklyn')
df['borough'].value_counts()

Manhattan        43792
Brooklyn         41843
Queens           13267
Bronx             2712
Staten Island      955
manhatan             1
Name: borough, dtype: int64

In [17]:
#replacing manhatan with Manhattan
df['borough'] = df['borough'].replace('manhatan', 'Manhattan')
df['borough'].value_counts()

Manhattan        43793
Brooklyn         41843
Queens           13267
Bronx             2712
Staten Island      955
Name: borough, dtype: int64

## 5. Data Consistency Checks

### 5a. Missing Data

In [18]:
#looking for missing data
df.isnull().sum().sort_values(ascending=False)

license                   102597
house_rules                52131
last_review                15893
reviews_per_month          15879
country                      532
availability_365             448
minimum_nights               409
review_rate_number           326
listings_per_host            319
host_identity_verified       289
service_fee                  273
listing_title                250
price                        247
construction_year            214
number_of_reviews            183
country_code                 131
instant_bookable             105
cancellation_policy           76
borough                       29
neighborhood                  16
longitude                      8
latitude                       8
id                             0
host_id                        0
room_type                      0
dtype: int64

In [19]:
# calculate the % missing values
percentage_result = df.isnull().sum()/df.shape[0] * 100
# create a DataFrame to show case the result
percentage_result = pd.DataFrame({"columns": percentage_result.keys(), "% missing": percentage_result})
percentage_result.reset_index(drop=True, inplace=True)
percentage_result

columns  % missing
0                       id   0.000000
1            listing_title   0.243667
2                  host_id   0.000000
3   host_identity_verified   0.281679
4                  borough   0.028265
5             neighborhood   0.015595
6                 latitude   0.007797
7                longitude   0.007797
8                  country   0.518524
9             country_code   0.127682
10        instant_bookable   0.102340
11     cancellation_policy   0.074075
12               room_type   0.000000
13       construction_year   0.208579
14                   price   0.240743
15             service_fee   0.266084
16          minimum_nights   0.398639
17       number_of_reviews   0.178364
18             last_review  15.490404
19       reviews_per_month  15.476759
20      review_rate_number   0.317742
21       listings_per_host   0.310919
22        availability_365   0.436651
23             house_rules  50.810437
24                 license  99.998051

In [20]:
#looking at the correlations between variables
df.corr()

id   host_id  latitude  longitude  \
id                  1.000000 -0.000830 -0.008832   0.042546   
host_id            -0.000830  1.000000  0.000661  -0.008999   
latitude           -0.008832  0.000661  1.000000   0.074348   
longitude           0.042546 -0.008999  0.074348   1.000000   
construction_year   0.001081  0.004871  0.005697   0.000861   
minimum_nights      0.005668 -0.002266  0.014889  -0.039639   
number_of_reviews  -0.041530 -0.004503 -0.025236   0.069169   
reviews_per_month   0.038038 -0.001746 -0.019492   0.118598   
review_rate_number  0.036633  0.003459 -0.003917   0.015283   
listings_per_host   0.024296  0.001722  0.032468  -0.104154   
availability_365   -0.139226 -0.002044 -0.005011   0.058428   

                    construction_year  minimum_nights  number_of_reviews  \
id                           0.001081        0.005668          -0.041530   
host_id                      0.004871       -0.002266          -0.004503   
latitude                     0.005697        0.014889          -0.025236   
longitude                    0.000861       -0.039639           0.069169   
construction_year            1.000000       -0.000486           0.001990   
minimum_nights              -0.000486        1.000000          -0.049997   
number_of_reviews            0.001990       -0.049997           1.000000   
reviews_per_month            0.004092       -0.096141           0.590939   
review_rate_number           0.004753       -0.002167          -0.018412   
listings_per_host           -0.002745        0.084846          -0.080907   
availability_365            -0.008264        0.063541           0.099368   

                    reviews_per_month  review_rate_number  listings_per_host  \
id                           0.038038            0.036633           0.024296   
host_id                     -0.001746            0.003459           0.001722   
latitude                    -0.019492           -0.003917           0.032468   
longitude                    0.118598            0.015283          -0.104154   
construction_year            0.004092            0.004753          -0.002745   
minimum_nights              -0.096141           -0.002167           0.084846   
number_of_reviews            0.590939           -0.018412          -0.080907   
reviews_per_month            1.000000            0.037526          -0.025621   
review_rate_number           0.037526            1.000000           0.024273   
listings_per_host           -0.025621            0.024273           1.000000   
availability_365             0.077193           -0.006217           0.159194   

                    availability_365  
id                         -0.139226  
host_id                    -0.002044  
latitude                   -0.005011  
longitude                   0.058428  
construction_year          -0.008264  
minimum_nights              0.063541  
number_of_reviews           0.099368  
reviews_per_month           0.077193  
review_rate_number         -0.006217  
listings_per_host           0.159194  
availability_365            1.000000

In [21]:
#looking at distinct values
df.nunique()

id                        102058
listing_title              61281
host_id                   102057
host_identity_verified         2
borough                        5
neighborhood                 224
latitude                   21991
longitude                  17774
country                        1
country_code                   1
instant_bookable               2
cancellation_policy            3
room_type                      4
construction_year             20
price                       1151
service_fee                  231
minimum_nights               153
number_of_reviews            476
last_review                 2477
reviews_per_month           1016
review_rate_number             5
listings_per_host             78
availability_365             438
house_rules                 1975
license                        1
dtype: int64

###### most of the values in 'id' and 'host_id' are unique which wont contribute to my analysis. 
###### country, country_code, and license only have 1 unique values and are not applicable to my analysis

In [22]:
#dropping unneeded columns from data as they are not applicable to the analysis
df.drop(columns = {'id'}, inplace = True)
df.drop(columns = {'host_id'}, inplace = True)
df.drop(columns = {'country_code'}, inplace = True)
df.drop(columns = {'country'}, inplace = True)
df.drop(columns = {'license'}, inplace = True)

In [23]:
#checking output
df.head()

listing_title host_identity_verified  \
0                Clean & quiet apt home by the park            unconfirmed   
1                             Skylit Midtown Castle               verified   
2               THE VILLAGE OF HARLEM....NEW YORK !                    NaN   
3                                               NaN            unconfirmed   
4  Entire Apt: Spacious Studio/Loft by central park               verified   

     borough  neighborhood  latitude  longitude instant_bookable  \
0   Brooklyn    Kensington  40.64749  -73.97237            False   
1  Manhattan       Midtown  40.75362  -73.98377            False   
2  Manhattan        Harlem  40.80902  -73.94190             True   
3   Brooklyn  Clinton Hill  40.68514  -73.95976             True   
4  Manhattan   East Harlem  40.79851  -73.94399            False   

  cancellation_policy        room_type  construction_year  price service_fee  \
0              strict     Private room             2020.0  $966        $193    
1            moderate  Entire home/apt             2007.0  $142         $28    
2            flexible     Private room             2005.0  $620        $124    
3            moderate  Entire home/apt             2005.0  $368         $74    
4            moderate  Entire home/apt             2009.0  $204         $41    

   minimum_nights  number_of_reviews last_review  reviews_per_month  \
0            10.0                9.0    10/19/21               0.21   
1            30.0               45.0     5/21/22               0.38   
2             3.0                0.0         NaN                NaN   
3            30.0              270.0      7/5/19               4.64   
4            10.0                9.0    11/19/18               0.10   

   review_rate_number  listings_per_host  availability_365  \
0                 4.0                6.0             286.0   
1                 4.0                2.0             228.0   
2                 5.0                1.0             352.0   
3                 4.0                1.0             322.0   
4                 3.0                1.0             289.0   

                                         house_rules  
0  Clean up and treat the home the way you'd like...  
1  Pet friendly but please confirm with me if the...  
2  I encourage you to use my kitchen, cooking and...  
3                                                NaN  
4  Please no smoking in the house, porch or on th...

In [24]:
#due to the high missing values in house_rules, it will be dropped
df.drop(columns = {'house_rules'}, inplace = True)

In [25]:
#checking output
df.head()

listing_title host_identity_verified  \
0                Clean & quiet apt home by the park            unconfirmed   
1                             Skylit Midtown Castle               verified   
2               THE VILLAGE OF HARLEM....NEW YORK !                    NaN   
3                                               NaN            unconfirmed   
4  Entire Apt: Spacious Studio/Loft by central park               verified   

     borough  neighborhood  latitude  longitude instant_bookable  \
0   Brooklyn    Kensington  40.64749  -73.97237            False   
1  Manhattan       Midtown  40.75362  -73.98377            False   
2  Manhattan        Harlem  40.80902  -73.94190             True   
3   Brooklyn  Clinton Hill  40.68514  -73.95976             True   
4  Manhattan   East Harlem  40.79851  -73.94399            False   

  cancellation_policy        room_type  construction_year  price service_fee  \
0              strict     Private room             2020.0  $966        $193    
1            moderate  Entire home/apt             2007.0  $142         $28    
2            flexible     Private room             2005.0  $620        $124    
3            moderate  Entire home/apt             2005.0  $368         $74    
4            moderate  Entire home/apt             2009.0  $204         $41    

   minimum_nights  number_of_reviews last_review  reviews_per_month  \
0            10.0                9.0    10/19/21               0.21   
1            30.0               45.0     5/21/22               0.38   
2             3.0                0.0         NaN                NaN   
3            30.0              270.0      7/5/19               4.64   
4            10.0                9.0    11/19/18               0.10   

   review_rate_number  listings_per_host  availability_365  
0                 4.0                6.0             286.0  
1                 4.0                2.0             228.0  
2                 5.0                1.0             352.0  
3                 4.0                1.0             322.0  
4                 3.0                1.0             289.0

In [26]:
#removing all missing values from dataset
print("Before dropping NaN rows:", df.shape)

df = df.dropna()
print("After dropping NaN rows:", df.shape)

Before dropping NaN rows: (102599, 19)
After dropping NaN rows: (84621, 19)


### 5b. Duplictes

In [27]:
#identifying duplicates
df[df.duplicated()]

listing_title  \
70825                    Penthouse Designer Loft Brooklyn   
71877                  Brooklyn, Clinton Hill, Private rm   
73083   Enormous and illuminated on top floor (long term)   
73098         Bright Modern Chic & Clean East Village APT   
73116                      Central Park - Room for 2 to 4   
...                                                   ...   
102588           Loft Space for Events, Meetings & Shoots   
102591                 Columbus Ave Apt 1 block from Park   
102593                Welcoming, Clean, Cheap on St Marks   
102595                      Best Location near Columbia U   
102597                   Big Studio-One Stop from Midtown   

       host_identity_verified    borough         neighborhood  latitude  \
70825             unconfirmed   Brooklyn          Sunset Park  40.63947   
71877             unconfirmed   Brooklyn         Clinton Hill  40.68405   
73083             unconfirmed     Queens             Elmhurst  40.74524   
73098             unconfirmed  Manhattan         East Village  40.72923   
73116                verified  Manhattan      Upper West Side  40.79701   
...                       ...        ...                  ...       ...   
102588            unconfirmed  Manhattan    Flatiron District  40.74068   
102591               verified  Manhattan      Upper West Side  40.77408   
102593               verified  Manhattan         East Village  40.72826   
102595            unconfirmed  Manhattan  Morningside Heights  40.80460   
102597            unconfirmed     Queens     Long Island City  40.74989   

        longitude instant_bookable cancellation_policy        room_type  \
70825   -74.01888            False            moderate  Entire home/apt   
71877   -73.96681            False            flexible     Private room   
73083   -73.88760            False            moderate  Entire home/apt   
73098   -73.98197            False            moderate  Entire home/apt   
73116   -73.96933             True            flexible     Private room   
...           ...              ...                 ...              ...   
102588  -73.98999             True            flexible  Entire home/apt   
102591  -73.98181            False              strict  Entire home/apt   
102593  -73.98422             True              strict     Private room   
102595  -73.96545             True            moderate     Private room   
102597  -73.93777             True              strict  Entire home/apt   

        construction_year    price service_fee  minimum_nights  \
70825              2007.0    $590        $118              2.0   
71877              2022.0  $1,146        $229              3.0   
73083              2010.0  $1,088        $218             60.0   
73098              2013.0    $732        $146              1.0   
73116              2019.0    $929        $186              1.0   
...                   ...      ...         ...             ...   
102588             2016.0    $618        $124              1.0   
102591             2005.0  $1,139        $228              5.0   
102593             2017.0  $1,099        $220              1.0   
102595             2016.0    $837        $167              1.0   
102597             2015.0    $546        $109              2.0   

        number_of_reviews last_review  reviews_per_month  review_rate_number  \
70825               220.0     6/27/19               3.10                 1.0   
71877                 2.0     9/14/15               0.04                 5.0   
73083                 2.0      1/2/19               0.17                 5.0   
73098                40.0     6/22/19               2.37                 4.0   
73116                27.0      6/9/19               1.69                 2.0   
...                   ...         ...                ...                 ...   
102588              177.0     6/29/19               3.78                 4.0   
102591               17.0      1/4/19               0.35                 5.0

In [28]:
#deleting duplicates
df.drop_duplicates(keep='first', inplace=True)

In [29]:
#checking shape
df.shape

(81566, 19)

In [30]:
#drop columns for listing title id since it's not appliable to the analysis
df.drop(columns = {'listing_title'}, inplace = True)

In [31]:
df.head()

host_identity_verified    borough        neighborhood  latitude  longitude  \
0            unconfirmed   Brooklyn          Kensington  40.64749  -73.97237   
1               verified  Manhattan             Midtown  40.75362  -73.98377   
4               verified  Manhattan         East Harlem  40.79851  -73.94399   
5               verified  Manhattan         Murray Hill  40.74767  -73.97500   
7            unconfirmed   Brooklyn  Bedford-Stuyvesant  40.68688  -73.95596   

  instant_bookable cancellation_policy        room_type  construction_year  \
0            False              strict     Private room             2020.0   
1            False            moderate  Entire home/apt             2007.0   
4            False            moderate  Entire home/apt             2009.0   
5             True            flexible  Entire home/apt             2013.0   
7            False            moderate     Private room             2009.0   

     price service_fee  minimum_nights  number_of_reviews last_review  \
0    $966        $193             10.0                9.0    10/19/21   
1    $142         $28             30.0               45.0     5/21/22   
4    $204         $41             10.0                9.0    11/19/18   
5    $577        $115              3.0               74.0     6/22/19   
7  $1,060        $212             45.0               49.0     10/5/17   

   reviews_per_month  review_rate_number  listings_per_host  availability_365  
0               0.21                 4.0                6.0             286.0  
1               0.38                 4.0                2.0             228.0  
4               0.10                 3.0                1.0             289.0  
5               0.59                 3.0                1.0             374.0  
7               0.40                 5.0                1.0             219.0

### 5c. Changing Data Types

In [32]:
#changing data type to datetime
df['last_review'] = pd.to_datetime(df['last_review'])

In [33]:
#changing price and service fee
df['price']=df['price'].apply(lambda x: x.replace('$','') if pd.isna(x) == False else x)
df['service_fee']=df['service_fee'].apply(lambda x: x.replace('$','') if pd.isna(x) == False else x)
df['service_fee']=df['service_fee'].astype(float)
df['price']=df['price'].apply(lambda x: x.replace(',','') if pd.isna(x) == False else x)
df['price']=df['price'].astype(float)

In [34]:
#changing data types
df['neighborhood'] = df['neighborhood'].astype('str')
df['borough'] = df['borough'].astype('str')
df['cancellation_policy'] = df['cancellation_policy'].astype('str')
df["number_of_reviews"] = np.array(df["number_of_reviews"], dtype = np.int16)
df["review_rate_number"] = np.array(df["review_rate_number"], dtype=np.int8)
df["listings_per_host"] = np.array(df["listings_per_host"], dtype = np.int16)

In [35]:
#checking output on data types
df.dtypes

host_identity_verified            object
borough                           object
neighborhood                      object
latitude                         float64
longitude                        float64
instant_bookable                  object
cancellation_policy               object
room_type                         object
construction_year                float64
price                            float64
service_fee                      float64
minimum_nights                   float64
number_of_reviews                  int16
last_review               datetime64[ns]
reviews_per_month                float64
review_rate_number                  int8
listings_per_host                  int16
availability_365                 float64
dtype: object

### 5d. Updating Incorrect Data

In [36]:
df['minimum_nights'].describe()

count    81566.000000
mean         7.471863
std         28.365003
min       -365.000000
25%          2.000000
50%          3.000000
75%          5.000000
max       5645.000000
Name: minimum_nights, dtype: float64

In [37]:
#checking the minimum night stay that shows less then 1
df[df['minimum_nights']<1]

host_identity_verified    borough        neighborhood  latitude  \
176              unconfirmed   Brooklyn         Fort Greene  40.69098   
398                 verified   Brooklyn          Kensington  40.64302   
421                 verified  Manhattan              Nolita  40.72094   
441                 verified  Manhattan              Harlem  40.80497   
478              unconfirmed  Manhattan     Upper West Side  40.77886   
525                 verified   Brooklyn  Bedford-Stuyvesant  40.68967   
42500            unconfirmed  Manhattan      Hell's Kitchen  40.76694   
42538               verified   Brooklyn  Bedford-Stuyvesant  40.68470   
69749               verified   Brooklyn        Williamsburg  40.71534   
91357               verified   Brooklyn             Gowanus  40.67070   

       longitude instant_bookable cancellation_policy        room_type  \
176    -73.97113            False            flexible     Private room   
398    -73.97255            False            flexible     Private room   
421    -73.99706            False            flexible  Entire home/apt   
441    -73.95016            False            moderate  Entire home/apt   
478    -73.98042             True              strict  Entire home/apt   
525    -73.95445            False            moderate  Entire home/apt   
42500  -73.98773             True            flexible  Entire home/apt   
42538  -73.94350             True            moderate  Entire home/apt   
69749  -73.94906            False            moderate     Private room   
91357  -73.99118             True            moderate      Shared room   

       construction_year   price  service_fee  minimum_nights  \
176               2008.0  1155.0        231.0           -10.0   
398               2015.0   779.0        156.0            -1.0   
421               2008.0   874.0        175.0           -10.0   
441               2021.0   920.0        184.0           -12.0   
478               2022.0   410.0         82.0            -2.0   
525               2022.0   413.0         83.0            -3.0   
42500             2009.0   711.0        142.0          -365.0   
42538             2022.0    85.0         17.0          -200.0   
69749             2012.0   328.0         66.0          -125.0   
91357             2014.0  1177.0        235.0           -10.0   

       number_of_reviews last_review  reviews_per_month  review_rate_number  \
176                  213  2019-06-24               2.00                   5   
398                   82  2019-05-19               0.94                   2   
421                   68  2019-06-10               0.69                   5   
441                  203  2019-07-06               2.14                   5   
478                  129  2019-06-07               1.33                   2   
525                   42  2019-05-18               0.44                   5   
42500                 13  2019-07-07               5.91                   4   
42538                  3  2019-04-23               1.06                   2   
69749                146  2019-06-23               1.78                   1   
91357                  9  2017-05-08               0.31                   2   

       listings_per_host  availability_365  
176                    2              19.0  
398                    2             131.0  
421                    1              13.0  
441                    3              77.0  
478                    1             381.0  
525                    1             292.0  
42500                  4               0.0  
42538                  1             157.0  
69749                  1              46.0  
91357                  1               0.0

In [38]:
#replacing with 0
df.loc[df['minimum_nights']<1,'minimum_nights']=1

In [39]:
#checking the minimum night stay that shows greater then 365
df[df['minimum_nights']>365]

host_identity_verified    borough        neighborhood  latitude  \
299              unconfirmed   Brooklyn          Greenpoint  40.73119   
473              unconfirmed  Manhattan  Washington Heights  40.84468   
1306             unconfirmed   Brooklyn            Bushwick  40.70202   
5768                verified  Manhattan   Greenwich Village  40.73293   
7356                verified     Queens    Long Island City  40.75104   
11194            unconfirmed   Brooklyn       Crown Heights  40.67255   
13405               verified  Manhattan              Harlem  40.82915   
26342            unconfirmed   Brooklyn        Williamsburg  40.71772   
42369            unconfirmed  Manhattan     Upper East Side  40.76174   
42398               verified   Brooklyn            Bushwick  40.70235   
42407            unconfirmed   Brooklyn           Bay Ridge  40.63189   
67608            unconfirmed  Manhattan     Upper East Side  40.77030   
69482            unconfirmed   Brooklyn            Bushwick  40.70202   
73027               verified   Brooklyn       Crown Heights  40.67255   
83506            unconfirmed   Brooklyn        Williamsburg  40.71772   
91347            unconfirmed     Queens           Rego Park  40.73048   
99691            unconfirmed  Manhattan              Harlem  40.81102   

       longitude instant_bookable cancellation_policy        room_type  \
299    -73.95578            False              strict     Private room   
473    -73.94303             True              strict     Private room   
1306   -73.92402            False              strict  Entire home/apt   
5768   -73.99782            False            moderate  Entire home/apt   
7356   -73.93863             True            moderate  Entire home/apt   
11194  -73.94914            False            moderate     Private room   
13405  -73.94034            False            moderate  Entire home/apt   
26342  -73.95059            False            moderate     Private room   
42369  -73.96625            False            moderate     Private room   
42398  -73.92892             True            flexible     Private room   
42407  -74.02322            False            moderate  Entire home/apt   
67608  -73.96115            False              strict  Entire home/apt   
69482  -73.92402            False              strict  Entire home/apt   
73027  -73.94914             True              strict     Private room   
83506  -73.95059             True              strict     Private room   
91347  -73.85331            False            moderate     Private room   
99691  -73.94712            False            moderate  Entire home/apt   

       construction_year   price  service_fee  minimum_nights  \
299               2011.0   614.0        123.0           452.0   
473               2016.0   409.0         82.0           398.0   
1306              2011.0  1179.0        236.0           370.0   
5768              2008.0   379.0         76.0          1250.0   
7356              2013.0   129.0         26.0           500.0   
11194             2022.0  1156.0        231.0           500.0   
13405             2016.0   805.0        161.0           999.0   
26342             2009.0   121.0         24.0           999.0   
42369             2016.0   360.0         72.0           954.0   
42398             2013.0   453.0         91.0          5645.0   
42407             2022.0   843.0        169.0          2645.0   
67608             2006.0   777.0        155.0           500.0   
69482             2011.0  1179.0        236.0           370.0   
73027             2022.0  1156.0        231.0           500.0   
83506             2009.0   121.0         24.0           999.0   
91347             2012.0   483.0         97.0           458.0   
99691             2015.0   303.0         61.0           825.0   

       number_of_reviews last_review  reviews_per_month  review_rate_number  \
299                    6  2016-11-13               0.10                   4   
473                   34  201

In [40]:
#replacing with 365
df.loc[df['minimum_nights']>365,'minimum_nights']=365

In [41]:
#identifying duplicates
df[df.duplicated()]

host_identity_verified   borough neighborhood  latitude  longitude  \
91357               verified  Brooklyn      Gowanus   40.6707  -73.99118   

      instant_bookable cancellation_policy    room_type  construction_year  \
91357             True            moderate  Shared room             2014.0   

        price  service_fee  minimum_nights  number_of_reviews last_review  \
91357  1177.0        235.0             1.0                  9  2017-05-08   

       reviews_per_month  review_rate_number  listings_per_host  \
91357               0.31                   2                  1   

       availability_365  
91357               0.0

In [42]:
#checking the availability_365 doesnt show any negative
df[df['availability_365']<0]

host_identity_verified    borough      neighborhood  latitude  \
490                 verified  Manhattan            Harlem  40.80224   
529                 verified  Manhattan    Hell's Kitchen  40.77090   
561                 verified   Brooklyn    Sheepshead Bay  40.58422   
621              unconfirmed     Queens  Long Island City  40.74581   
683                 verified   Brooklyn        Greenpoint  40.73731   
...                      ...        ...               ...       ...   
86504               verified  Manhattan    Hell's Kitchen  40.76045   
86555               verified   Brooklyn          Flatbush  40.65251   
86675               verified      Bronx          Longwood  40.81937   
86735            unconfirmed  Manhattan      East Village  40.72627   
86764               verified     Queens         Ridgewood  40.70971   

       longitude instant_bookable cancellation_policy        room_type  \
490    -73.94558             True            moderate     Private room   
529    -73.99181            False            moderate     Private room   
561    -73.94079             True            moderate  Entire home/apt   
621    -73.95295             True            moderate     Private room   
683    -73.95450            False              strict     Private room   
...          ...              ...                 ...              ...   
86504  -73.99810             True            moderate  Entire home/apt   
86555  -73.96325             True            moderate     Private room   
86675  -73.90978            False            moderate     Private room   
86735  -73.99145             True            moderate  Entire home/apt   
86764  -73.89741            False            flexible  Entire home/apt   

       construction_year   price  service_fee  minimum_nights  \
490               2016.0   424.0         85.0             2.0   
529               2021.0   619.0        124.0            28.0   
561               2007.0   441.0         88.0             4.0   
621               2016.0  1112.0        222.0            30.0   
683               2016.0   232.0         46.0             3.0   
...                  ...     ...          ...             ...   
86504             2017.0   471.0         94.0            30.0   
86555             2007.0   158.0         32.0             4.0   
86675             2005.0   614.0        123.0             2.0   
86735             2005.0   335.0         67.0             3.0   
86764             2009.0   606.0        121.0            28.0   

       number_of_reviews last_review  reviews_per_month  review_rate_number  \
490                   17  2015-10-01               0.18                   3   
529                   43  2019-06-29               0.45                   5   
561                  128  2016-11-15               1.34                   2   
621                   65  2017-07-31               0.74                   1   
683                    4  2015-12-04               0.05                   2   
...                  ...         ...                ...                 ...   
86504                  1  2017-04-20               0.04                   4   
86555                  1  2017-07-03               0.04                   5   
86675                 68  2019-05-30               1.83                   2   
86735                 27  2019-04-23               0.76                   5   
86764                  1  2018-09-15               0.10                   4   

       listings_per_host  availability_365  
490                    2              -9.0  
529                    2              -4.0  
561                    2              -8.0  
621                    3              -6.0  
683                   28              -2.0  
...                  ...               ...  
86504                121              -2.0  
86555                  1              -6.0  
86675                  1              -7.0  
86735                  1              -4.0  
86764                  1              -

In [43]:
#replacing with 0
df.loc[df['availability_365']<0,'availability_365']=0

In [44]:
#checking the availabilty isnt greater than 365
df[df['availability_365']>365]

host_identity_verified    borough        neighborhood  latitude  \
5                   verified  Manhattan         Murray Hill  40.74767   
9                unconfirmed  Manhattan     Upper West Side  40.80178   
110                 verified   Brooklyn             Gowanus  40.68157   
130              unconfirmed  Manhattan              Harlem  40.81618   
131                 verified   Brooklyn        Clinton Hill  40.68414   
...                      ...        ...                 ...       ...   
86849               verified   Brooklyn  Bedford-Stuyvesant  40.68285   
86857               verified  Manhattan     Upper West Side  40.79415   
86858            unconfirmed  Manhattan              Harlem  40.81589   
86874            unconfirmed  Manhattan     Upper West Side  40.76986   
86879               verified  Manhattan     Lower East Side  40.71429   

       longitude instant_bookable cancellation_policy        room_type  \
5      -73.97500             True            flexible  Entire home/apt   
9      -73.96723            False              strict     Private room   
110    -73.98989            False              strict  Entire home/apt   
130    -73.94894             True            flexible  Entire home/apt   
131    -73.96351            False            moderate     Private room   
...          ...              ...                 ...              ...   
86849  -73.91484             True            flexible     Private room   
86857  -73.97323             True            flexible     Private room   
86858  -73.94648             True            flexible  Entire home/apt   
86874  -73.98673             True              strict     Private room   
86879  -73.98755             True            moderate     Private room   

       construction_year  price  service_fee  minimum_nights  \
5                 2013.0  577.0        115.0             3.0   
9                 2015.0  291.0         58.0             2.0   
110               2011.0  528.0        106.0            30.0   
130               2005.0  449.0         90.0             3.0   
131               2018.0  930.0        186.0             3.0   
...                  ...    ...          ...             ...   
86849             2022.0  543.0        109.0            14.0   
86857             2019.0  969.0        194.0             5.0   
86858             2003.0  749.0        150.0             4.0   
86874             2018.0  643.0        129.0             1.0   
86879             2007.0  759.0        152.0             2.0   

       number_of_reviews last_review  reviews_per_month  review_rate_number  \
5                     74  2019-06-22               0.59                   3   
9                    118  2017-07-21               0.99                   5   
110                   19  2017-03-17               0.20                   2   
130                  155  2019-06-20               1.42                   3   
131                  260  2019-07-03               2.35                   4   
...                  ...         ...                ...                 ...   
86849                 32  2019-01-02               0.92                   4   
86857                  3  2019-06-13               0.08                   2   
86858                125  2019-06-20               3.37                   4   
86874                  1  2016-08-09               0.03                   5   
86879                 14  2017-07-28               0.38                   1   

       listings_per_host  availability_365  
5                      1             374.0  
9                      1             375.0  
110                    1             400.0  
130                    3             410.0  
131                    1             388.0  
...                  ...               ...  
86849                  4             370.0  
86857                  1             380.0  
86858                  4             409.0  
86874                  1             383.0  
86879                  1   

In [45]:
#replacing with 365
df.loc[df['availability_365']>365,'availability_365']=365

In [46]:
#checking final shape
df.shape

(81566, 18)

In [47]:
#checking total missing values
(df.isnull().sum()).sum()

0

In [48]:
#identifying duplicates
df[df.duplicated()]

host_identity_verified    borough neighborhood  latitude  longitude  \
70936               verified  Manhattan  Murray Hill  40.74776  -73.97335   
72821            unconfirmed   Brooklyn  Sunset Park  40.64685  -74.01097   
91357               verified   Brooklyn      Gowanus  40.67070  -73.99118   

      instant_bookable cancellation_policy        room_type  \
70936             True            flexible  Entire home/apt   
72821            False            moderate  Entire home/apt   
91357             True            moderate      Shared room   

       construction_year   price  service_fee  minimum_nights  \
70936             2014.0    96.0         19.0             4.0   
72821             2022.0   618.0        124.0             3.0   
91357             2014.0  1177.0        235.0             1.0   

       number_of_reviews last_review  reviews_per_month  review_rate_number  \
70936                  7  2015-09-04               0.10                   2   
72821                 34  2019-06-16               0.75                   2   
91357                  9  2017-05-08               0.31                   2   

       listings_per_host  availability_365  
70936                  1               0.0  
72821                  1               0.0  
91357                  1               0.0

In [49]:
#deleting duplicates
df.drop_duplicates(keep='first', inplace=True)

In [50]:
#checking final shape
df.shape

(81563, 18)

## 6. Exporting Data

In [51]:
#exporting data
df.to_csv(os.path.join(path, '02 Data','Prepared Data', 'NYC_cleaned_data.csv'))